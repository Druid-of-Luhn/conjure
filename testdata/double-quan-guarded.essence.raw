Spec
  { language = "Essence"
  , version = [ 2
              , 0
              ]
  , topLevelBindings = [ ( Letting
                         , "myForall"
                         , DeclQuantifier (
                             DeclLambda
                               [ ( "__myForall_x"
                                 , TypeBoolean
                                 )
                               , ( "__myForall_y"
                                 , TypeBoolean
                                 )
                               ] (
                               GenericNode
                                 And
                                 [ Identifier "__myForall_x"
                                 , Identifier "__myForall_y"
                                 ] ) ) (
                             DeclLambda
                               [ ( "__myForall_x"
                                 , TypeBoolean
                                 )
                               , ( "__myForall_y"
                                 , TypeBoolean
                                 )
                               ] (
                               GenericNode
                                 Imply
                                 [ Identifier "__myForall_x"
                                 , Identifier "__myForall_y"
                                 ] ) ) (
                             ValueBoolean True )
                         )
                       , ( Find
                         , "s"
                         , DomainSet
                             { size = Nothing
                             , minSize = Nothing
                             , maxSize = Nothing
                             , attrDontCare = False
                             , element = DomainIntegerFromTo (
                                           Just ( ValueInteger 1 ) ) (
                                           Just ( ValueInteger 90 ) )
                             , representation = Nothing
                             }
                         )
                       ]
  , topLevelWheres = []
  , objective = Nothing
  , constraints = [ ExprQuantifier
                      { quanName = Identifier "myForall"
                      , quanVar = Identifier "i"
                      , quanOver = DomainIntegerFromTo (
                                     Just ( ValueInteger 1 ) ) (
                                     Just ( ValueInteger 9 ) )
                      , quanGuard = Nothing
                      , quanBody = ExprQuantifier
                                     { quanName = Identifier "myForall"
                                     , quanVar = Identifier "j"
                                     , quanOver = DomainIntegerFromTo (
                                                    Just ( ValueInteger 1 ) ) (
                                                    Just ( ValueInteger 9 ) )
                                     , quanGuard = Just (
                                                     GenericNode
                                                       Eq
                                                       [ Identifier "i"
                                                       , GenericNode
                                                           Times
                                                           [ ValueInteger 2
                                                           , Identifier "j"
                                                           ]
                                                       ] )
                                     , quanBody = GenericNode
                                                    Elem
                                                    [ GenericNode
                                                        Plus
                                                        [ Identifier "i"
                                                        , Identifier "j"
                                                        ]
                                                    , Identifier "s"
                                                    ]
                                     }
                      }
                  ]
  , metadata = []
  }