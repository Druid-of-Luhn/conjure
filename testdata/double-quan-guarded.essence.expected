language Essence 2.0

letting myForall be quantifier { append { x : bool, y : bool -> x /\ y }
                                 guard { x : bool, y : bool -> x => y }
                                 identity true
                               }
find s : set of int(1..90)

such that
    myForall i : int(1..9) . myForall j : int(1..9) , i = 2 * j . i + j elem s
